### ISSUE TITLE
pre-proposal: CurveZMQ

### ISSUE THREAD

**Issue by minrk** [jupyter/enhancement-proposals#75]
zeromq has a transport-level encryption and authentication protocol called [CurveZMQ][].

[CurveZMQ]: https://rfc.zeromq.org/spec/26/

I've just landed support for CurveZMQ in [IPython Parallel][ipp-curve], and I think it's worth talking about in Jupyter.

[ipp-curve]: https://github.com/ipython/ipyparallel/pull/553

The gist of the most basic implementation:

- **server** sockets set a public and private key, issued e.g. via `zmq.curve_keypair()` from pyzmq. (the sockets that bind make the most sense, but is not strictly a requirement), and `CURVE_SERVER=1` to enable auth.
- **client** sockets use the server's public key as a **server key** - anyone with this key can connect to the server and send/receive messages
- client's own private/public key can be any key pair, and are used exclusively for encryption; they have no role in authentication

In Jupyter, we already have a key distribution mechanism, which is the HMAC message-signing key in connection files. We can use the same key distribution for Curve keys. In the context of Jupyter, it's a little weird, because it's usually the client (KernelManager) that *sets* the credentials, which means the client issues the kernel's private key as well, and needs to pass the *private* key to the kernel. This being the case, the absolute simplest version is to use the same private/public keypair for both ends.

### Sketch:

- KernelManager issues private, public keypair with `zmq.curve_keypair()`
- store private_key and public_key in connection file
- kernel sockets set CURVE_PRIVATEKEY, CURVE_PUBLICKEY, CURVE_SERVER=1
- client sockets set the same CURVE_PRIVATEKEY, CURVE_PUBLICKEY, and use the public key in CURVE_SERVERKEY (alternative: client sockets issue new private/public keys, but if the private key is already in the connection file, I see no benefit to issuing more keypairs)

Here's an example of an authenticated socket pair in pyzmq:

<details>

<summary>curve socket example</summary>

```python
import asyncio
import zmq
import zmq.asyncio as zaio

async def main():
    public, private = zmq.curve_keypair()
    ctx = zaio.Context()
    server = ctx.socket(zmq.ROUTER)
    # server socket is a 'curve server'
    server.CURVE_SECRETKEY = private
    server.CURVE_PUBLICKEY = public
    server.CURVE_SERVER = True

    url = "tcp://127.0.0.1:5555"
    server.bind(url)

    no_auth_client = ctx.socket(zmq.DEALER)

    auth_client = ctx.socket(zmq.DEALER)
    auth_client.CURVE_SECRETKEY = private
    auth_client.CURVE_PUBLICKEY = public
    auth_client.CURVE_SERVERKEY = public # this authenticates the client

    auth_client.connect(url)
    no_auth_client.connect(url)


    for i in range(5):
        # messages from 'auth_client' will be received
        asyncio.ensure_future(auth_client.send(b'auth'))
        # messages from 'no_auth_client' will never be delivered
        asyncio.ensure_future(no_auth_client.send(b'noauth'))
        msg = await server.recv_multipart()
        print("Received", msg)

    ctx.destroy(linger=0)

if __name__ == "__main__":
    asyncio.run(main())
```

</details>

### Benefits

- transport-level encryption, which was not previously available. CurveZMQ provides forward-secrecy, meaning access to the keys is *not* sufficient to decrypt captured traffic in the future. It is sufficient for a live man-in-the-middle attack.
- connection-level authentication removes need to check message signatures (the less security-related code we implement ourselves, the better!), and is vastly simpler to implement for both kernels and clients
- connection-level authentication prevents non-authenticated clients with access to ports from monitoring IOPub, which is only currently preventable when using `ipc://` transport and file permissions

### Caveats

- Severe crash issues in libzmq regarding threadsafety on systems without `getrandom`, closed as "wontfix: use getrandom": https://github.com/zeromq/libzmq/issues/4241 . I think this is a huge mistake, and have backported the opt-in patches on libzmq to both pyzmq's bundled libzmq and conda-forge libzmq, but may turn out to be a major problem.
- Requires implementation by kernels, and capability advertisement in kernelspecs; always a high hurdle

### Alternatives:

- broader support for zmq auth, including PLAIN (username:password, authentication but not encryption), GSSAPI, etc.
- full support for ZAP (zmq auth protocol), which is a much bigger burden on kernels, but would enable more sophisticated access control

Both of these would require more generic exposure of general options for zmq, whereas the proposal as it is only requires sharing of a single string (or two strings) for the key pair, under the exact same model we already have, and only setting a 3 socket options (same values on all sockets), a fairly minor change in practice. Plus, all changes are in socket creation, nowhere else in the protocol implementation.

**Comment by minrk** [comment]
we can avoid passing the private key in the connection file by specifying that it will be in an environment variable, e.g. `$JUPYTER_CURVE_SECRETKEY`. In which case, clients should assume that they will need to issue their own private/public keypair. This doesn't really make any difference to clients, but it means the connection file is no longer enough info for a man-in-the-middle attack. It's also no longer enough info for kernels to set up their sockets.

**Comment by krassowski** [comment]
@minrk if we were to implement this in 2026, would any of the above plan need to change? I can spend time on implementing a PoC for it this week, just wanted to check if there were any developments, especially on ZMQ side that would impact this pre-proposal.

**Comment by minrk** [comment]
No, I don't think anything relevant has changed, I think this plan should still work.

**Comment by jasongrout** [comment]
@krassowski - https://github.com/jupyter-server/jupyter_server/pull/1638#discussion_r3211126393 indicates you deviated from the plan above in the implementation. When you have a few minutes, could you circle back here to discuss how your implementation differed from the plan above?

**Comment by krassowski** [comment]
@jasongrout I re-aligned the same day (see commits [`jupyter_client@68a485a`](https://github.com/jupyter/jupyter_client/pull/1110/commits/68a485a0d16056dc31194b28af1b6ce4d361adfd), [`jupyter_server@e16e61a`](https://github.com/jupyter-server/jupyter_server/pull/1638/commits/e16e61aa35b2dbc667df0a2549ce9f0cd774af6d), [`ipykernel@8d56ecc`](https://github.com/ipython/ipykernel/pull/1515/commits/8d56ecc4d31ef6f1bec9bf746d0f716c796ce8e8)) - there is no deviation that I aware of. Sorry about the confusion.

**Comment by westurner** [comment]
Original question: what would need to change later in this for PQ; 

"Implementing Post-Quantum CurveZMQ Support" someday https://gemini.google.com/share/338fec880d38 ;

> [ libsodium, libzmq, pyzmq, jupyter ]

> Yes, unequivocally. If your goal is to secure Jupyter and ZeroMQ traffic with X25519MLKEM768 today, bypassing libzmq’s native CurveZMQ mechanism is dramatically easier, safer, and faster to implement

> Instead, you can achieve PQ security immediately by treating ZeroMQ strictly as a local messaging layer and pushing the cryptography to a TLS 1.3 Proxy layer that already supports hybrid ML-KEM

> ...

> The architecture you are describing already has a strong precedent in projects like `jupyter_kernel_gateway` and Jupyter Enterprise Gateway (JEG)

- jupyter-server/kernel_gateway: https://github.com/jupyter-server/kernel_gateway
- jupyter-server/enterprise_gateway: https://github.com/jupyter-server/enterprise_gateway

> In a traditional setup, the Jupyter Server acts as a translator: it talks WebSockets to the browser frontend and ZeroMQ to the local backend kernel. Enterprise Gateway severs that ZMQ tie. It exposes a purely HTTP/WebSocket API. When you run code, the frontend sends a WebSocket message to the Gateway, which then manages the remote kernels.  
>
> By pushing the communication standard to WebSockets, enterprise deployments can drop a standard ingress proxy (like Envoy or Nginx) in front of the Gateway. You can enable X25519MLKEM768 on that proxy today, and instantly achieve a post-quantum secure edge without modifying a single line of Jupyter's source code

**Comment by krassowski** [comment]
@minrk do you think you would have time to review the PRs with PoC any time soon? Would it be ok to publish a set of pre-releases with encryption included to allow for evaluation of this functionality downstream?

**Comment by minrk** [comment]
Yes, I will definitely put some time into reviewing them next week, thanks for putting them together!

**Comment by Carreau** [comment]
This is now mostly implemented in Jupyter_client and ipykernel.

**Comment by JohanMabille** [comment]
I think it could be worth opening a JEP so that it become part of the protocol and other kernels are encouraged to implement it.
